text/0000-set-helper.md

```diff
@@ -0,0 +1,106 @@
+- Start Date: 2020-02-20
+- Relevant Team(s): Ember.js
+- RFC PR: (after opening the RFC PR, update this with a link to it and update the file name)
+- Tracking: (leave this empty)
+
+# `set` helper
+
+## Summary
+
+This RFC introduces a new `set` helper like the one provided by
+[`ember-simple-set-helper`](https://github.com/pzuraq/ember-simple-set-helper).
+
+The `set` helper will always return a function that will change a value.
+
+```hbs
+<SelectCountry @update={{set this.country}} />
+```
+
+The new `set` helper can be used instead of the old `{{fn (mut ..)}}` combination.
+However this RFC does not deprecate `{{mut}}`.
+
+## Motivation
+
+The existing `mut` helper is quite helpful, but behaves often in an unexpected way.
+
+When used to pass an argument as in `<MyComponent @foo={{mut this.bar}}>` or
+`{{my-component foo=(mut this.bar)}}` it will do nothing for a glimmer component
+but provide a box with a `value` and an `update` function on the private `attrs`
+property of a classic component.
+
+However the more common use-case is to pass the result of `mut` to either the 
+`{{action}}` or the `{{fn}}` helper:
+
+```hbs
+<SelectCountry @update={{fn (mut this.country)}} />
+<button {{on "click" (fn (mut this.name) "")}}>delete name</button>
+```
+
+However many people would expect `{{fn (mut this.foo)}}` to be the same as
+`{{mut this.foo}}` in all cases, which can lead to confusion.
+
+So this proposal proposes a new `set` helper that will work as `mut` wrapped
+inside `fn`.
+
+## Detailed design
+
+The new `set` helper can be called with one or two arguments.
+
+### one argument version
+
+The one argument version returns a function that receives one argument `value` and
+will set the property passed as argument to `value`.
+```hbs
+{{set context.property}}
+```
+
+```js
+return function(arg) {
+  context.property = arg;
+};
+```
+
+The first argument must *always* have a context. So `{{set foo}}` or `{{set @foo}}` is not valid.
+This aligns with [RFC 0308](https://github.com/emberjs/rfcs/blob/master/text/0308-deprecate-property-lookup-fallback.md).
+
+The fact that the `set` helper somehow needs to know the *context* and the *name* of the
+argument it receives makes it magic syntax that can not be used by a userland helper without
+a AST transformation. But the need for the helper seems big enough to justify this, especially
+when the same magic syntax is already used by the `mut` helper.
```

Comment:
What about this scenario?

```hbs
{{#each @values as |value|}}
  <FancyInput @onChange={{set value}} />
{{/each}}
```

This seems like a *very* common use case, and one that the current proposed design would rule out.

Reply:
Perhaps this is part of the reason this doesn't fully replace `{{mut}}`! If so, that's probably worth making explicit in the text of the RFC.

Reply:
I just realized that this exact behavior probably doesn't work *at all* regardless of this. A better example would be something like:

```hbs
{{#each @users as |user|}}
  <FancyInput @onChange={{set user.name}} />
{{/each}}
```

The alternative, less magical version as suggested in another comment would just be:

```hbs
{{#each @users as |user|}}
  <FancyInput @onChange={{set user 'name'}} />
{{/each}}
```

I could see an argument either way here in terms of the brevity-vs.-magic tradeoff. Apologies for the confusion!

Reply:
exactly. `{{set user.name}}` would work. Also this does *not* work currently (just tested it):

```
{{#each @values as |value|}}
  <FancyInput @onChange={{fn (mut value)}} />
{{/each}}
```

I generally would want `{{set user.name}}` and not `{{set user "name"}}`. But indeed, both would work.